lib/Service/DocumentService.php

```diff
@@ -182,59 +192,68 @@ public function get($documentId) {
 	 * @param $version
 	 * @return array
 	 * @throws DoesNotExistException
-	 * @throws VersionMismatchException
+	 * @throws InvalidArgumentException
 	 */
 	public function addStep($documentId, $sessionId, $steps, $version): array {
-		$document = null;
-		$oldVersion = null;
-
-		try {
-			$this->lockingProvider->acquireLock('document-push-lock-' . $documentId, ILockingProvider::LOCK_EXCLUSIVE);
-		} catch (LockedException $e) {
-			throw new VersionMismatchException('Version does not match');
+		$stepsToInsert = [];
+		$querySteps = [];
+		$getStepsSinceVersion = null;
+		$newVersion = $version;
+		foreach ($steps as $step) {
+			if ($step > "AAE") {
+				array_push($stepsToInsert, $step);
+			} else {
+				array_push($querySteps, $step);
+			}
 		}
+		if (sizeof($stepsToInsert) > 0) {
+			$newVersion = $this->insertSteps($documentId, $sessionId, $stepsToInsert, $version);
+		}
+		// If there were any queries in the steps send the entire history
+		$getStepsSinceVersion = sizeof($querySteps) > 0 ? 0 : $version;
+		return [
+			'steps' => $this->getSteps($documentId, $getStepsSinceVersion),
+			'version' => $newVersion
+		];
+	}
 
+	/**
+	 * @param $documentId
+	 * @param $sessionId
+	 * @param $steps
+	 * @param $version
+	 * @return int
+	 * @throws DoesNotExistException
+	 * @throws InvalidArgumentException
+	 */
+	private function insertSteps($documentId, $sessionId, $steps, $version): int {
+		$document = null;
+		$stepsVersion = null;
 		try {
 			$document = $this->documentMapper->find($documentId);
-			if ($version !== $document->getCurrentVersion()) {
-				throw new VersionMismatchException('Version does not match');
-			}
 			$stepsJson = json_encode($steps);
 			if (!is_array($steps) || $stepsJson === null) {
 				throw new InvalidArgumentException('Failed to encode steps');
 			}
-			$validStepTypes = ['addMark', 'attr', 'removeMark', 'replace', 'replaceAround'];
-			foreach ($steps as $step) {
-				if (array_key_exists('stepType', $step) && !in_array($step['stepType'], $validStepTypes, true)) {
-					throw new InvalidArgumentException('Invalid step data');
-				}
-			}
-			$oldVersion = $document->getCurrentVersion();
-			$newVersion = $oldVersion + count($steps);
+			$stepsVersion = $this->stepMapper->getLatestVersion($document->getId());
+			$newVersion = $stepsVersion + count($steps);
```

Comment:
If we get rid of the lock we may want to wrap this in a transaction so that reading the current version, increasing and writing steps is an atomic operation.

Should be easy with https://docs.nextcloud.com/server/latest/developer_manual/basics/storage/database.html#transactions

Reply:
I created a follow up issue in #3726